### PR TITLE
[#156][Enhancement] Feature to read password from user rather than from HDFS file

### LINKED ISSUE #156: [Connector] [JDBC] Feature to read password from user rather than from HDFS file
**What feature are your requesting?**

Read password from user rather than from HDFS file

**What benefits do you foresee from the feature you are requesting?**

Currently, users can put JDBC source password in an HDFS file and the Gimel connector reads the password from there. This causes a tight coupling with HDFS.

**Potential solution/ideas?**

Read password directly from user

**Additional context**

Refer `gimel.jdbc.p.file` here: https://github.com/paypal/gimel/blob/master/docs/gimel-connectors/teradata.md#teradata-read-options

### PR DESCRIPTION
**Pull request by rajarshisarkar**
Make sure you have checked all steps below.

### GitHub Issue
Fixes #156 

### Checklist:

- [x] This pull request updates the documentation
- [ ] This pull request changes library dependencies
- [x] Title of the PR is of format (example) : [#25][Github] Add Pull Request Template

### Commit Guidelines
- [x] My commits all reference GH issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by rajarshisarkar** [comment]
@Dee-Pac @bthilakraj @laxpatil Can you please review the same.

**Comment by Dee-Pac** [comment]
Hi @rajarshisarkar Thanks for the PR, let me take a look and get back.

**Comment by laxpatil** [comment]
@rajarshisarkar : Thanks for your PR.

Major concern about adding password as part of sparkConf parameters is it will be visible in Spark UI(RM) . in system properties since it is passed as Spark Conf parameters. Any spark user can see the this property in Spark UI. It could be a security concern.

@Dee-Pac LMK if you have any inputs on this and if you would like to add anything.

**Comment by rajarshisarkar** [comment]
@laxpatil Thanks for your inputs. Do let me know your thoughts in case the user encrypts and passes the password. In that way, we can decrypt the password internally and it won't be visible in the Spark UI too.

File: gimel-dataapi/gimel-connectors/gimel-jdbc/src/main/scala/com/paypal/gimel/jdbc/utilities/JDBCAuthUtilities.scala [hunk 1]
```diff
@@ -78,40 +78,64 @@ class JDBCAuthUtilities(sparkSession: SparkSession) extends Serializable {
...
   }
 
+  /**
+    * This function returns the Gimel JDBC password
+    *
+    * @param dataSetProps dataset properties
+    * @return Gimel JDBC password
+    */
+  private def getPassword(dataSetProps: Map[String, Any]): String = {
+    dataSetProps.getOrElse(JdbcConfigs.jdbcPassword, new String()).toString
```
Comment (Dee-Pac) [inline gimel-dataapi/gimel-connectors/gimel-jdbc/src/main/scala/com/paypal/gimel/jdbc/utilities/JDBCAuthUtilities.scala]: @rajarshisarkar Please change this logic to throw exception if the `JdbcConfigs.jdbcPassword` is not present in the properties. Exception could be along the lines ... `gimel.jdbc.p.strategy=inline was supplied, but gimel.jdbc.password is NOT set.`

Comment (Dee-Pac) [reply gimel-dataapi/gimel-connectors/gimel-jdbc/src/main/scala/com/paypal/gimel/jdbc/utilities/JDBCAuthUtilities.scala]: @rajarshisarkar Also, if the flow enters this method - please print a warning `gimel.jdbc.p.strategy=inline is NOT a secure way to supply password. Please switch to gimel.jdbc.p.strategy=file !`

File: gimel-dataapi/gimel-connectors/gimel-jdbc/src/main/scala/com/paypal/gimel/jdbc/utilities/JDBCAuthUtilities.scala [hunk 2]
```diff
@@ -78,40 +78,64 @@ class JDBCAuthUtilities(sparkSession: SparkSession) extends Serializable {
...
+        }
+        if (!flag) {
+          val msg =
+            """Username or Password NOT FOUND!
+              |Check the configuration parameter or password file"""".stripMargin
+          throw new JDBCAuthException(msg, null)
+        }
+        (userName, password)
+      }
+      case "input" => {
```
Comment (Dee-Pac) [inline gimel-dataapi/gimel-connectors/gimel-jdbc/src/main/scala/com/paypal/gimel/jdbc/utilities/JDBCAuthUtilities.scala]: this block can be replaced as 
```scala
case "inline" => (userName,getPassword(dataSetProps))
```

File: gimel-dataapi/gimel-connectors/gimel-jdbc/src/main/scala/com/paypal/gimel/jdbc/utilities/JDBCCommons.scala [hunk 3]
```diff
@@ -146,6 +146,9 @@ object JDBCCommons {
           case "file" => {
             // do nothing
           }
+          case "input" => {
```
Comment (Dee-Pac) [inline gimel-dataapi/gimel-connectors/gimel-jdbc/src/main/scala/com/paypal/gimel/jdbc/utilities/JDBCCommons.scala]: please rename to `inline`

File: docs/gimel-connectors/teradata.md [hunk 4]
```diff
@@ -253,8 +256,8 @@ val options = Map(
           ("numPartitions", s"${partitions}")
         , ("fetchSize", s"${fetchSize}")
         , ("gimel.jdbc.read.type", "FASTEXPORT")
-        ,("gimel.jdbc.p.strategy","file")
-        ,("gimel.jdbc.p.file","hdfs://path")
+        ,("gimel.jdbc.p.strategy","input")
```
Comment (Dee-Pac) [inline docs/gimel-connectors/teradata.md]: Could you please add another example ? let's retain the existing documentation for file.

**Comment by Dee-Pac** [comment]
@laxpatil You're right, we should absolutely avoid passing passwords inline. But at the same time,  I think we should leave that choice to user since passing inline is mostly useful for debugging. 

Also, since the password is not passed at the beginning of spark session as `spark.*` it will not be visible in the Spark UI. But mostly be retained in logs or say, jupyter notebooks. 

Hence, I've requested @rajarshisarkar to print a warning to user that this should not be used for production deployments.

File: gimel-dataapi/gimel-connectors/gimel-jdbc/src/main/scala/com/paypal/gimel/jdbc/utilities/JDBCAuthUtilities.scala [hunk 5]
```diff
@@ -78,40 +78,64 @@ class JDBCAuthUtilities(sparkSession: SparkSession) extends Serializable {
     dataSetProps.getOrElse(JdbcConfigs.jdbcP, defaultValue).toString
   }
 
+  /**
+    * This function returns the Gimel JDBC password
+    *
+    * @param dataSetProps dataset properties
+    * @return Gimel JDBC password
+    */
+  private def getPassword(dataSetProps: Map[String, Any]): String = {
```
Comment (Dee-Pac) [inline gimel-dataapi/gimel-connectors/gimel-jdbc/src/main/scala/com/paypal/gimel/jdbc/utilities/JDBCAuthUtilities.scala]: Could you please rename this explicitly as `getInlinePassword`?

**Comment by rajarshisarkar** [comment]
@Dee-Pac Implemented the review comments. Please check.

**Pull request merged**